### PR TITLE
Add PaddlePaddle backend for einops

### PR DESCRIPTION
**Pull request by zhouwei25**
Add PaddlePaddle backend for einops

### CODE REVIEW

**Comment by zhouwei25** [comment]
@arogozhnikov hi，I want to ask how to trick CI?

**Comment by arogozhnikov** [comment]
@zhouwei25 you need to add paddle in test installation https://github.com/arogozhnikov/einops/blob/master/test.py#L44

**Comment by zhouwei25** [comment]
@arogozhnikov Hello, We are PaddlePaddle official Team, our team want to support paddle backend for einops. Thank you to  review and give advice to us!

File: einops/_backends.py [hunk 1]
```diff
@@ -614,3 +614,70 @@ def layers(self):
...
+        if len(axes) == x.ndim:
+            return super().reduce(x, operation, axes).squeeze(0)
+        else:
+            return super().reduce(x, operation, axes)
+
+    def transpose(self, x, axes):
+        return x.transpose(axes)
+
+    def add_axes(self, x, n_axes, pos2len):
+        repeats = [1] * n_axes
```
Comment (JunnYu) [inline einops/_backends.py:651]: repeats = [-1] * n_axes

File: einops/_backends.py [hunk 2]
```diff
@@ -614,3 +614,70 @@ def layers(self):
...
+
+class PaddleBackend(AbstractBackend):
+    framework_name = "paddle"
+
+    def __init__(self):
+        import paddle
+        self.paddle = paddle
+
+    def is_appropriate_type(self, tensor):
+        return isinstance(tensor, self.paddle.Tensor, self.paddle.static.Variable)
```
Comment (JunnYu) [inline einops/_backends.py:627]: return isinstance(tensor, (self.paddle.Tensor, self.paddle.static.Variable)), missing  "()"

File: einops/_backends.py (hunk 1 above)
Comment (arogozhnikov) [reply einops/_backends.py:651]: +1

File: einops/_backends.py (hunk 2 above)
Comment (arogozhnikov) [reply einops/_backends.py:627]: +1

**Comment by arogozhnikov** [comment]
after fixing above comments by YunnYu, I still see a large number of crashes on existing tests. 

Some of them are shape related, like
```
assert x.shape == (2, 3, 4)
```
I'd recommend adding comparison with tuple to `__eq__`, as other frameworks do.

Other wide class of errors are .reshape([])  (reshape to zero-dim tensor).

Probably there are more test failures, but these are most common.

File: einops/_backends.py (hunk 2 above)
Comment (zhouwei25) [reply einops/_backends.py]: done, thx~

**Comment by zhouwei25** [comment]
> after fixing above comments by YunnYu, I still see a large number of crashes on existing tests.
> 
> Some of them are shape related, like
> 
> ```
> assert x.shape == (2, 3, 4)
> ```
> 
> I'd recommend adding comparison with tuple to `__eq__`, as other frameworks do.
> 
> Other wide class of errors are .reshape([]) (reshape to zero-dim tensor).
> 
> Probably there are more test failures, but these are most common.

Hi, `assert out_array.shape == out.shape` occurres this problem.  I fix it to assert `tuple(out_array.shape)` , because paddle framework shape is python list type, cannot overwrite __eq__ like torch. I think it doesn't matter, because tuple/list type are same in use most time.

**Comment by zhouwei25** [comment]
I tests in my machine locally, and these is not other problem. All other problem is :

**Comment by arogozhnikov** [comment]
@zhouwei25 started CI, see error messages above, they seem similar to what I get on local setup with x86 + python 3.10

**Comment by zhouwei25** [comment]
> Details

These problems all have been supported in paddlepaddle latest nightly version. Only paddlepaddle==0.0.0 support this. But if I install paddlepaddle==0.0.0, because it also update protobuf to 4.2.0 in `requirement.txt` , which is not support in tensorflow and will result unittests failed in tensorflow. Do you have any suggestions?

**Comment by zhouwei25** [comment]
> @zhouwei25 started CI, see error messages, they seem similar to what I get on local setup with x86 + python 3.10

@arogozhnikov  These problems all have been supported in paddlepaddle latest nightly version. Only paddlepaddle==0.0.0 support this. But if I install paddlepaddle==0.0.0, because it also update protobuf to 4.2.0 in requirement.txt , which is not support in tensorflow and will result unittests failed in tensorflow. Do you have any suggestions? Thank you.

**Comment by zhouwei25** [comment]
@arogozhnikov hi, do you have some suggestions?

**Comment by arogozhnikov** [comment]
One way would be to create a separate entrance point that will test specifically paddle, and not other frameworks. Not that it is great solution, but it will work. I'll hate maintaining two test points, it makes hard to test things on user side.

More wisely (from your side) would be not depending on `protobuf` package, instead shipping it as a subpackage like `paddle.protobuf`. Another option is ship it as protobuf4 or protobuf-paddle, so it could be installed independently from protobuf package.

Protobuf is a bad dependency now because some packages require new version, and some require old (already happened). For users having TF and paddle at the same time should be ok IMO, but that's currently impossible.

I highly recommend you folks to take steps and untie paddle from protobuf package, this will help you and your users.

**Comment by zhouwei25** [comment]
@arogozhnikov hi, I just install paddle but import failed, may be machine environment problem? do you know the problem?
infoflow 2023-03-07 14-34-58

**Comment by arogozhnikov** [comment]
`run(...)` runs bash script, not python. Do smth like `run("python -c 'import paddle' ")`

**Comment by zhouwei25** [comment]
> `run(...)` runs bash script, not python. Do something like `run("python -c 'import paddle' ")`

Thank you. There is a solution. tensorflow install protobuf==3.19.6 by default, but its master branch use > 3.20, so it will upgrade in future. Paddle install protobuf==4.20.2 by default,  but >=3.20.2 is enough. So we just need to limit protobuf to >=3.20.2, and it's compatible for different framework.

CI and UT has passed. Hope to merge and add paddle backend.

**Comment by zhouwei25** [comment]
@JunnYu @arogozhnikov hello, CI has passed, help review and merge, thx

**Comment by zhouwei25** [comment]
@arogozhnikov @JunnYu  hi, CI has passed. can you help to review and merge the PR? thank you

**Comment by arogozhnikov** [comment]
> we just need to limit protobuf to >=3.20.2, and it's compatible for different framework.

This will not be for long (at least I can't rely on that).

I'll need to rewrite CI first to have separate testing envs for frameworks. Hopefully will have time this week

**Comment by zhouwei25** [comment]
> > we just need to limit protobuf to >=3.20.2, and it's compatible for different framework.
> 
> This will not be for long (at least I can't rely on that).
> 
> I'll need to rewrite CI first to have separate testing envs for frameworks. Hopefully will have time this week

Hi, thanks, I think einops can add an requirements.txt to limit its own requirements, not only be desined by framework. 
Otherwise, separate diffentent unittest Framework. Paddle will Release before April 5. We hope to join the einops In this release.  Can it support in that time? thank you very much.

**Comment by arogozhnikov** [comment]
@zhouwei25 testing was rewritten, PR merged.

Version with paddle support can be installed with 
```bash
pip install git+https://github.com/arogozhnikov/einops.git
```

Planning release in the next two weeks, you should be good.

**Pull request merged**